### PR TITLE
Rate limit warning refactor

### PR DESCRIPTION
**Pull request by Raemon**
This PR does several things:
– add the rate limit warning to PostsNewForm and PostsEditForm
– refactor the rate limit warning code, such that `rateLimitDateWhenUserNextAbleToPost` and `rateLimitDateWhenUserNextAbleToComment` each end up looking at all the currently applying rate limits, and selecting the longest (strictest) one, and providing a bit of messaging to the user.
– removes the `postSpecificRateLimit` field from post schema, and incorporates it into the user `userNextAbleToComment` field

Instead of throwing errors, all rate limits now return a RateLimitInto object which includes rateLimitType (for analytics), nextEligible (to post/comment) and rateLimitMessage.

This ended up being a surprisingly gnarly refactor. I think the resulting code makes more sense and can handle a wide variety of rate-limiting experiments without unintuitive errors and edge-cases.

image┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204542982045019) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/editor/RateLimitWarning.tsx [hunk 1]
```diff
@@ -3,20 +3,22 @@ import { Components, registerComponent } from '../../lib/vulcan-lib';
...
+  // particular copyπ that our product team wants. In the future we probably want to pass in the reason
   // that the user is currently rate-limited, to display the appropriate message.
   if (!isEAForum && !lastRateLimitExpiry) return null
 
   const diffInMin = moment(lastRateLimitExpiry).diff(moment(), 'minutes')
   const fromNow = moment(lastRateLimitExpiry).fromNow()
-  const message = isEAForum ?
-    `You've written more than 3 comments in the last 30 min. Please wait ${diffInMin} min before commenting again. You'll be able to post more comments as your karma increases.` :
-    `Please wait ${fromNow} before commenting again.`
+  const eaForumMessage = `You've written more than 3 comments in the last 30 min. Please wait ${diffInMin} min before commenting again. You'll be able to post more comments as your karma increases.`
```
Comment (Raemon) [inline packages/lesswrong/components/editor/RateLimitWarning.tsx:18]: EA forum folk, you might want to add the rateLimit message here but didn't want to surprise you with it.

File: packages/lesswrong/components/editor/RateLimitWarning.tsx [hunk 2]
```diff
@@ -3,21 +3,23 @@ import { Components, registerComponent } from '../../lib/vulcan-lib';
...
+// Tells the user when they can next comment or post if they're rate limited, and a brief explanation
+const RateLimitWarning = ({lastRateLimitExpiry, rateLimitMessage}: {
   lastRateLimitExpiry?: Date|null,
+  rateLimitMessage?: string|null
 }) => {
   // Sorry this is not great. Basically, I wanted to keep the previous functionality for the default case
   // (which shows how long you have until you can comment again), and then on the EA Forum, show the
   // particular copy that our product team wants. In the future we probably want to pass in the reason
   // that the user is currently rate-limited, to display the appropriate message.
-  if (!isEAForum && !lastRateLimitExpiry) return null
```
Comment (Raemon) [inline packages/lesswrong/components/editor/RateLimitWarning.tsx]: I'm quite confused about what's going on here – seems like either lastRateLimitExpiry should be required (and this null-return handled in the parent component) or I'd expect EAForum to also want to return null here. What's going on with this chesterton fence?

Comment (s-cheng) [reply packages/lesswrong/components/editor/RateLimitWarning.tsx]: Sorry, the copy changed after I wrote that comment. You can see the latest [here](https://github.com/ForumMagnum/ForumMagnum/pull/7128/files#diff-e7d7df813ced31a065a267f489d6e4b228ebc2054aa3059c9fc6a08e10141fd5)

**Review by b0b3rt — APPROVED** [review]

File: packages/lesswrong/server/callbacks/rateLimits.ts [hunk 3]
```diff
@@ -103,34 +102,13 @@ export const getNthMostRecentItemDate = async function<
...
-
-    const postsInPastTimeframe = await userNumberOfItemsInPastTimeframe(user, Posts, hours)
-  
-    if (postsInPastTimeframe > 0) {
-      throw new Error(MODERATOR_ACTION_TYPES[moderatorRateLimit.type]);
+  const rateLimit = await rateLimitDateWhenUserNextAbleToPost(user);
+  if (rateLimit) {
+    const {nextEligible} = rateLimit;
+    if (nextEligible > new Date()) {
+      throw new Error(`Rate limit: You cannot comment until ${nextEligible}`);
```
Comment (b0b3rt) [inline packages/lesswrong/server/callbacks/rateLimits.ts:109]: You cannot post (not comment)

**Comment by Raemon** [comment]
Test cases (make sure the rate limit is enforced, and that we get a reasonable looking rate limit warning)

- User cannot post twice in 8 seconds
- User cannot post more than 5 times in 24 hours
- If a user is given a moderator rate limit, they cannot post more than once per time-period

- User cannot comment twice in 8 (15?) seconds
- If a forum hasn't set a downvote ratio, they don't get a downvoteRatio rate limit
- If a forum _has_ set a downvote ratio of 30%, if they get one downvoted, then they can't make more than 4 comments in 30 minutes 
- if a forum doesn't have a lowKarma threshold, users can make  5+ comments without restriction
- if a forum has a lowKarma threshold and users are below threshold, they can't make more than 4 comments per 30 minutes. (need to test this separate from the downvote ratio)

(both lowKarma and downvoteRatio should check and restrict comments on all posts whether or not the commenter owns those posts)


– if a user has a mod-rate-limit, they can't comment more than once per period on other people's posts (but can still comment once even if they've commented on their own post in the same period) (this applies on both EA Forum and LW atm, because the mod-rate-limits are all very strict, and I'm pretty sure EA forum wouldn't want to include comments on their own posts)
---- mod-rate-limited-user comments on their own post, and then can comment immediately on someone else's post
---- mod-rate-limited-user comments on someone else's post, immediately cannot comment again on that or any other user's post
---- most-rate-limited-user comments on someone else's post, can still comment on their own post
---- if we impose a rate limit after a user has commented recently, they should be unable to comment until [postedAt date of last comment + rate-limit-interval]


– if a user has the post-specific-rate-limit, they can't comment more than 3 times on any given post unless the post is their own, or has the ignoreRateLimits flag

**Pull request merged**